[lldb][test] Toolchain detection rewrite in Python

dzhidzhoev opened the pull request:
This fix is based on a problem with cxx_compiler and cxx_linker macros
on Windows.
There was an issue with compiler detection in paths containing "icc". In
such case, Makefile.rules thought it was provided with icc compiler.

To solve that, utilities detection has been rewritten in Python.
The last element of compiler's path is separated, taking into account
the platform path delimiter, and compiler type is extracted, with regard
of possible cross-toolchain prefix.

dzhidzhoev committed:
[lldb][test] Improve toolchain detection in Makefile.rules

This fix is based on a problem with cxx_compiler and cxx_linker macros
on Windows.
There was an issue with compiler detection in paths containing "icc". In
such case, Makefile.rules thought it was provided with icc compiler.

To solve that, utilities detection has been rewritten in Python.
The last element of compiler's path is separated, taking into account
platform path delimiter, and compiler type is extracted, with regard
of possible cross-toolchain prefix.  Paths for additional tools,
like OBJCOPY, are initialized with tools built with LLVM if USE_LLVM_TOOLS
is on, to achieve better portability for Windows.
Quotes from paths are removed in Makefile.rules, that may be added when
arguments are passed from Python to make.

labath commented:
Umm... could we move this logic to python, and then just pass the final tool paths (or whatever) into the makefile, like we did with the other patches (for `(HOST_)OS`, etc.) ?

labath commented:
(that said, I would be very surprised if anyone was running this test suite with icc, so I think ripping out `icc` support is also an option)

dzhidzhoev commented:
> Umm... could we move this logic to python, and then just pass the final tool paths (or whatever) into the makefile, like we did with the other patches (for `(HOST_)OS`, etc.) ?

Moved logic to Python, updated commit description.

dzhidzhoev commented:
Updated added dwp tool handling, to adopt changes from e77ac42bccb8c26bbf4b74d8e92eb09e7fa1b218 and resolve merge conflict.

dzhidzhoev committed:
Fix comment

dzhidzhoev commented:
Now this is ready for review.

labath commented on lldb/packages/Python/lldbsuite/test/builders/builder.py:
```diff
@@ -96,16 +98,120 @@ def getArchSpec(self, architecture):
+
+            cc = cc.strip()
+            cc_path = pathlib.Path(cc)
+            cc = cc_path.as_posix()
```
What's up with the posix thing? Could we stick to native paths?

dzhidzhoev replied:
The idea in Makefile.rules was to unify path separators to process them. Indeed, we don't need that here.

labath commented on lldb/packages/Python/lldbsuite/test/builders/builder.py:
```diff
@@ -96,16 +98,120 @@ def getArchSpec(self, architecture):
+
+            def getLlvmUtil(util_name):
+                llvm_tools_dir = os.getenv("LLVM_TOOLS_DIR", cc_dir.as_posix())
+                return llvm_tools_dir + "/" + util_name + exe_ext
```
```suggestion
                return os.path.join(llvm_tools_dir, util_name + exe_ext)
```

dzhidzhoev replied:
Fixed

labath commented on lldb/packages/Python/lldbsuite/test/make/Makefile.rules:
```diff
@@ -102,15 +102,33 @@ endif
 $(error "C compiler is not specified. Please run tests through lldb-dotest or lit")
 endif
 
+# Remove all " and ' characters from the path. Also remove surrounding space chars.
+cstrip = $(strip $(subst ",,$(subst ',,$(1))))
+override CC := $(call cstrip,$(CC))
+override CCC := $(call cstrip,$(CCC))
+override CXX := $(call cstrip,$(CXX))
+# Always override the linker. Assign already normalized CC.
+override LD := $(call cstrip,$(CC))
+# A kind of linker. It always gets retrieved from CC.
+override LDC := $(call cstrip,$(CCC))
+override OBJCOPY := $(call cstrip,$(OBJCOPY))
+override STRIP := $(call cstrip,$(STRIP))
+override ARCHIVER := $(call cstrip,$(ARCHIVER))
+override AR := $(call cstrip,$(AR))
+override DWP := $(call cstrip,$(DWP))
+override LLVM_AR := $(call cstrip,$(LLVM_AR))
```
Is this really necessary? Could we get rid of it by *not* passing the extra quotes in the python cmd?

dzhidzhoev replied:
Yes, you're right, fixed it.

labath commented on lldb/packages/Python/lldbsuite/test/builders/builder.py:
```diff
@@ -96,16 +98,120 @@ def getArchSpec(self, architecture):
+                utils.extend(['AR="%slibtool"' % os.getenv("CROSS_COMPILE", "")])
+
+            return [
+                'CC="%s"' % cc,
+                'CCC="%s"' % cc_type,
+                'CXX="%s"' % cxx,
```
I.e., by replacing this (and others) with
```suggestion
                'CC=%s' % cc,
                'CCC=%s' % cc_type,
                'CXX=%s' % cxx,
```

labath commented on lldb/packages/Python/lldbsuite/test/builders/builder.py:
```diff
@@ -96,16 +98,120 @@ def getArchSpec(self, architecture):
+                utils.extend(['LLVM_AR="%s"' % llvm_ar])
+
+            if not lldbplatformutil.platformIsDarwin():
+                if os.getenv("USE_LLVM_TOOLS"):
```
I don't see any reference to this var in the llvm tree. Who sets this? (and why)

labath replied:
I'm still puzzled by this..

dzhidzhoev replied:
I got rid of environment variable usage and used dotest.py param instead.

The flag `--use-llvm-tools` flag is added to force usage of LLVM tools passed with `--llvm-tools-dir` instead of compiler-provided tools.
We use it while running API tests on Windows x86-host/Ubuntu AArch64-target combination with cl.exe. This flag enables switching to the "local" tools usage without relying on the PATH/compiler path for utilities lookup, so as not to rely on an external toolchain while having all necessary tools built together with lldb.

labath replied:
Ok, I see now. This is new functionality added in this patch. Could we put that in a separate PR and keep this one as a NFC-ish refactoring/cleanup patch? I'd like to discuss this concept further, but I think it would be easier to do it in isolation.

dzhidzhoev replied:
Sorry for the delay, I had to take a leave.
Removed that from this PR.

labath replied:
Thanks. Before you go on the next PR, let me just elaborate on what I'm thinking.

Basically, the (API) test suite can run in two modes. The first one (the default) is where it uses all of the tools from the build tree that it can. This includes the compiler and the other tools like llvm-objcopy, etc. This is done for two reasons:
- it minimizes the amount of external dependencies
- it maximizes the reproducibility of the build

This mode doesn't (shouldn't?) require any special flags or configuration. The second mode is where you can run the tests against a toolchain (mostly: compiler) of your choosing, and there it's expected that you may need to pass many arguments in order to make it work.

The thing which bothers me about this flag is that it's a flag that forces a switch back to the default mode (for some tools), which is makes reasoning about it more complicated. I think it would be better to just make the behavior you want to be the default, and then (if necessary) have a separate flag to force the usage of a specific tool (e.g. objcopy). I think that would be similar to how things work with e.g. dsymutil, where by default we pick the in-tree version, but we have a cmake/dotest/etc. flag to let the user choose a different one.

This would mean that users that run tests against gcc (and which want to continue using the gcc's version of objcopy) could no longer rely on the compiler-based detection and would have to pass (and add it to dotest first) an additional flag explicitly, but I think that's fine, because:
- These tests mostly don't care about the objcopy version. Their main purpose is to capture the differences in DWARF emitted by different compilers. We have better ways of testing the differences in object file (elf) output.
- I'm not sure if we have any such users at the moment.

dzhidzhoev replied:
> The thing which bothers me about this flag is that it's a flag that forces a switch back to the default mode (for some tools), which is makes reasoning about it more complicated. I think it would be better to just make the behavior you want to be the default, and then (if necessary) have a separate flag to force the usage of a specific tool (e.g. objcopy). I think that would be similar to how things work with e.g. dsymutil, where by default we pick the in-tree version, but we have a cmake/dotest/etc. flag to let the user choose a different one.

It looks good for me. I'll do it this way.

labath commented on lldb/packages/Python/lldbsuite/test/make/Makefile.rules:
```diff
@@ -213,7 +229,7 @@ endif
 NO_LIMIT_DEBUG_INFO_FLAGS =
 MODULE_DEBUG_INFO_FLAGS =
-ifneq (,$(findstring clang,$(CC)))
+ifeq ($(CCC), clang)
```
What does CCC stand for?

labath replied:
So the last `C` stands for "type" ? :)

Can we call it CC_TYPE/KIND (or similar) instead? It's not like we're using it that often...

dzhidzhoev replied:
Renamed it

dzhidzhoev committed:
Typo fix

Co-authored-by: Pavel Labath <pavel@labath.sk>

dzhidzhoev committed:
Don't quote paths

dzhidzhoev committed:
Join paths with os.path.join

dzhidzhoev committed:
Use native paths

labath commented on lldb/packages/Python/lldbsuite/test/builders/builder.py:
```diff
@@ -96,16 +98,119 @@ def getArchSpec(self, architecture):
         if not cc and configuration.compiler:
             cc = configuration.compiler
+
         if cc:
```
```suggestion
        if not cc:
          return []
```

.. and then unindent the rest of the code

dzhidzhoev replied:
Done.

labath commented on lldb/packages/Python/lldbsuite/test/builders/builder.py:
```diff
@@ -96,16 +98,119 @@ def getArchSpec(self, architecture):
+
+            util_names = {
+                "OBJCOPY": "objcopy",
+                "STRIP": "objcopy",
```
Can we get rid of this variable? I don't think it's very helpful. Even though `objcopy` supports (I think) all of the functionality of `strip`, their args are different, so I think this could be pretty confusing.

dzhidzhoev replied:
Sorry, it was just a copy-paste typo.

labath commented on lldb/packages/Python/lldbsuite/test/builders/builder.py:
```diff
@@ -96,16 +98,119 @@ def getArchSpec(self, architecture):
+            utils = []
+
+            # Note: LLVM_AR is currently required by API TestBSDArchives.py tests.
+            # Assembly a full path to llvm-ar for given LLVM_TOOLS_DIR;
```
```suggestion
            # Assemble a full path to llvm-ar for given LLVM_TOOLS_DIR;
```

dzhidzhoev replied:
Fixed

labath reviewed (commented):
Looks pretty good now, but I still have some questions.

dzhidzhoev committed:
Use cmd arguments instead of env variables in builder

dzhidzhoev committed:
Rename CCC to CC_TYPE

dzhidzhoev committed:
Fix typo

dzhidzhoev committed:
Fix wrong utility for STRIP

dzhidzhoev committed:
Reformat condition

dzhidzhoev committed:
Formatting fixed

dzhidzhoev committed:
Revert "Use cmd arguments instead of env variables in builder"

This reverts commit 55b981cb0fcb4c2d3b62d8412b507afedd8801da.

dzhidzhoev committed:
Remove USE_LLVM_TOOLS usage

labath commented on lldb/packages/Python/lldbsuite/test/builders/builder.py:
```diff
@@ -96,17 +98,105 @@ def getArchSpec(self, architecture):
+        if not cc:
+            return []
+
+        exe_ext = ""
+        if lldbplatformutil.getHostPlatform() == "windows":
+            exe_ext = ".exe"
+
```
It looks like this isn't used now.

dzhidzhoev replied:
Fixed.

labath reviewed (approved):
Looks good now, just remove the dead code.

dzhidzhoev committed:
Removed unused exe_ext variable
